src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,114 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int MIN_ARRAY_SIZE = 10;
+    private Object[] arr;
+    private int size;
+
+    public ArrayList() {
+        this.arr = new Object[MIN_ARRAY_SIZE];
+        this.size = 0;
+    }
+
     @Override
     public void add(T value) {
-
+        ensureCapacity();
+        arr[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index >= 0 && index <= size) {
+            ensureCapacity();
+            System.arraycopy(arr, index, arr, index + 1, size - index);
+            arr[index] = value;
+            size++;
+        } else {
+            throw new ArrayListIndexOutOfBoundsException("Index out of bounds");
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        ensureCapacity(list.size());
+        for (int i = 0; i < list.size(); i++) {
+            arr[size++] = list.get(i);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index >= 0 && index < size) {
+            return (T) arr[index];
+        } else {
+            throw new ArrayListIndexOutOfBoundsException("Index out of bounds");
```

Comment:
The exception message 'Index out of bounds' is not very informative. Consider including the invalid index and the valid range in the message to make it more informative.